File: apps/mobile/AIActive/backend/utils/config_crypto.py
"""
配置文件加密/解密工具
与 Android 端保持一致的算法实现

使用 AES-256-GCM 加密，PBKDF2 从版本号派生密钥

@author Claude Code
@since 2026-02-06
"""
import base64
import hashlib
from Crypto.Cipher import AES
from Crypto.Protocol.KDF import PBKDF2
from Crypto.Random import get_random_bytes

ALGORITHM = 'AES'
SALT = b'JC_AI_CONFIG_SALT_2026'
ITERATIONS = 100000  # 测试用100K，生产环境使用300K
KEY_SIZE = 32  # 256 bits
GCM_TAG_LENGTH = 16  # 128 bits
GCM_IV_LENGTH = 12  # 96 bits


def derive_key(version: str) -> bytes:
    """
    从版本号派生加密密钥
    v1.2.0 -> "120" -> PBKDF2-HMAC-SHA256 -> 256-bit key

    安全说明：
    - 使用 PBKDF2-HMAC-SHA256 进行密钥派生
    - 迭代次数设置为 300,000 以防止暴力破解
    - 版本号标准化：右填充 '0' 到至少3位 (v1.0 -> "100", v1.2.0 -> "120")

    :param version: 版本号字符串 (e.g., "v1.2.0")
    :return: 256-bit AES 密钥
    """
    # 转换版本号为数字字符串: "v1.2.0" -> "120"
    # 与 Android 端保持一致：右填充 '0' 到至少 3 位
    version_num = version.replace('v', '').replace('.', '')
    while len(version_num) < 3:
        version_num = version_num + '0'

    # 使用 HMAC-SHA256，与 Android 端保持一致
    from Crypto.Hash import SHA256
    key = PBKDF2(
        version_num,
        SALT,
        dkLen=KEY_SIZE,
        count=ITERATIONS,
        hmac_hash_module=SHA256
    )
    return key


def encrypt(plaintext: str, version: str) -> str:
    """
    加密明文字符串

    :param plaintext: 要加密的明文
    :param version: 版本号（用于派生密钥）
    :return: Base64 编码的密文 (IV + ciphertext + auth_tag)
    """
    key = derive_key(version)
    iv = get_random_bytes(GCM_IV_LENGTH)

    cipher = AES.new(key, AES.MODE_GCM, nonce=iv)
    ciphertext, tag = cipher.encrypt_and_digest(plaintext.encode('utf-8'))

    # 组合 IV + ciphertext + tag
    combined = iv + ciphertext + tag

    # 返回 Base64 编码
    return base64.b64encode(combined).decode('utf-8')


def decrypt(ciphertext: str, version: str) -> str:
    """
    解密密文字符串

    :param ciphertext: Base64 编码的密文 (IV + ciphertext + auth_tag)
    :param version: 版本号（用于派生密钥）
    :return: 解密后的明文
    :raises ValueError: 如果解密失败（认证失败）
    """
    key = derive_key(version)

    # 解码 Base64
    combined = base64.b64decode(ciphertext)

    # 提取组件
    iv = combined[:GCM_IV_LENGTH]
    tag = combined[-GCM_TAG_LENGTH:]
    encrypted = combined[GCM_IV_LENGTH:-GCM_TAG_LENGTH]

    cipher = AES.new(key, AES.MODE_GCM, nonce=iv)
    plaintext = cipher.decrypt_and_verify(encrypted, tag)

    return plaintext.decode('utf-8')


def encrypt_dict(config: dict, version: str, keys_to_encrypt: list = None) -> dict:
    """
    加密字典中的指定字段

    :param config: 配置字典
    :param version: 版本号
    :param keys_to_encrypt: 要加密的键列表，None 表示加密所有字符串值
    :return: 加密后的字典
    """
    encrypted = {}

    for key, value in config.items():
        if keys_to_encrypt is None or key in keys_to_encrypt:
            if isinstance(value, str):
                encrypted[key] = encrypt(value, version)
            elif isinstance(value, dict):
                encrypted[key] = encrypt_dict(value, version, keys_to_encrypt)
            else:
                encrypted[key] = value
        else:
            encrypted[key] = value

    return encrypted


def decrypt_dict(config: dict, version: str, keys_to_decrypt: list = None) -> dict:
    """
    解密字典中的指定字段

    :param config: 加密的配置字典
    :param version: 版本号
    :param keys_to_decrypt: 要解密的键列表，None 表示解密所有字符串值
    :return: 解密后的字典
    """
    decrypted = {}

    for key, value in config.items():
        if keys_to_decrypt is None or key in keys_to_decrypt:
            if isinstance(value, str):
                try:
                    # 检查是否为服务器额外加密
                    if value.startswith("SERVER_ENC:"):
                        # 服务器额外加密，暂不解密
                        decrypted[key] = value
                    else:
                        decrypted[key] = decrypt(value, version)
                except Exception as e:
                    print(f"Warning: Failed to decrypt {key}: {e}")
                    decrypted[key] = value
            elif isinstance(value, dict):
                decrypted[key] = decrypt_dict(value, version, keys_to_decrypt)
            else:
                decrypted[key] = value
        else:
            decrypted[key] = value

    return decrypted


if __name__ == '__main__':
    # 测试加密/解密
    test_version = "v1.0.0"
    test_plaintext = "test_api_key_12345"

    print(f"Testing encryption/decryption with version: {test_version}")
    print(f"Plaintext: {test_plaintext}")

    # 加密
    encrypted = encrypt(test_plaintext, test_version)
    print(f"Encrypted: {encrypted}")

    # 解密
    decrypted = decrypt(encrypted, test_version)
    print(f"Decrypted: {decrypted}")

    # 验证
    assert decrypted == test_plaintext, f"Decryption failed! Expected {test_plaintext}, got {decrypted}"
    print("[OK] Encryption/decryption test passed!")

    # 测试字典加密
    test_config = {
        "api_key": "secret_key_123",
        "secret": "another_secret",
        "url": "https://example.com"
    }

    encrypted_config = encrypt_dict(test_config, test_version, ["api_key", "secret"])
    print(f"\nEncrypted config: {encrypted_config}")

    decrypted_config = decrypt_dict(encrypted_config, test_version, ["api_key", "secret"])
    print(f"Decrypted config: {decrypted_config}")

    assert decrypted_config["api_key"] == test_config["api_key"]
    assert decrypted_config["secret"] == test_config["secret"]
    assert decrypted_config["url"] == test_config["url"]
    print("[OK] Dictionary encryption/decryption test passed!")

File: apps/mobile/AIActive/android/app/src/main/java/com/jcoding/aiactivity/utils/ConfigCrypto.java
package com.jcoding.aiactivity.utils;

import android.util.Base64;
import android.util.Log;

import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.security.spec.KeySpec;
import java.util.Arrays;

/**
 * 配置文件加密/解密工具
 * - 基础层：AES-256-GCM，使用版本号派生密钥
 * - 增强层：服务器额外加密的高度敏感字段
 *
 * @author Claude Code
 * @since 2026-02-06
 */
public class ConfigCrypto {

    private static final String TAG = "ConfigCrypto";
    private static final String ALGORITHM = "AES";
    private static final String TRANSFORMATION = "AES/GCM/NoPadding";
    private static final String KEY_DERIVATION = "PBKDF2WithHmacSHA256";
    private static final String SALT = "JC_AI_CONFIG_SALT_2026";
    private static final int ITERATIONS = 100000;  // 测试用100K，生产环境使用300K
    private static final int KEY_SIZE = 256;
    private static final int GCM_TAG_LENGTH = 128;
    private static final int GCM_IV_LENGTH = 12;

    /**
     * 从版本号派生加密密钥
     * v1.2.0 -> "120" -> PBKDF2 -> 256-bit key
     *
     * 安全说明：
     * - 使用 PBKDF2WithHmacSHA256 进行密钥派生
     * - 迭代次数设置为 300,000 以防止暴力破解
     * - 版本号标准化：右填充 '0' 到至少3位 (v1.0 -> "100", v1.2.0 -> "120")
     *
     * @param version 版本号字符串 (e.g., "v1.2.0")
     * @return 256-bit AES 密钥
     * @throws Exception 如果密钥派生失败
     */
    public static SecretKey deriveKey(String version) throws Exception {
        // 转换版本号为数字字符串: "v1.2.0" -> "120"
        // 与 Python 端保持一致：右填充 '0' 到至少 3 位
        String versionNum = version.replace("v", "").replace(".", "");
        while (versionNum.length() < 3) {
            versionNum = versionNum + "0";
        }

        KeySpec spec = new PBEKeySpec(
            versionNum.toCharArray(),
            SALT.getBytes(StandardCharsets.UTF_8),
            ITERATIONS,
            KEY_SIZE
        );

        SecretKeyFactory factory = SecretKeyFactory.getInstance(KEY_DERIVATION);
        byte[] keyBytes = factory.generateSecret(spec).getEncoded();
        return new SecretKeySpec(keyBytes, ALGORITHM);
    }

    /**
     * 加密明文字符串
     *
     * @param plaintext 要加密的明文
     * @param version 版本号（用于派生密钥）
     * @return Base64 编码的密文 (IV + ciphertext + auth_tag)
     * @throws Exception 如果加密失败
     */
    public static String encrypt(String plaintext, String version) throws Exception {
        SecretKey key = deriveKey(version);

        // 生成随机 IV
        byte[] iv = new byte[GCM_IV_LENGTH];
        new SecureRandom().nextBytes(iv);

        // 初始化加密器
        Cipher cipher = Cipher.getInstance(TRANSFORMATION);
        GCMParameterSpec parameterSpec = new GCMParameterSpec(GCM_TAG_LENGTH, iv);
        cipher.init(Cipher.ENCRYPT_MODE, key, parameterSpec);

        // 加密
        byte[] ciphertext = cipher.doFinal(plaintext.getBytes(StandardCharsets.UTF_8));

        // 组合 IV + ciphertext
        byte[] combined = new byte[GCM_IV_LENGTH + ciphertext.length];
        System.arraycopy(iv, 0, combined, 0, GCM_IV_LENGTH);
        System.arraycopy(ciphertext, 0, combined, GCM_IV_LENGTH, ciphertext.length);

        // 返回 Base64 编码
        return Base64.encodeToString(combined, Base64.NO_WRAP);
    }

    /**
     * 解密密文字符串
     *
     * @param ciphertext Base64 编码的密文 (IV + ciphertext + auth_tag)
     * @param version 版本号（用于派生密钥）
     * @return 解密后的明文
     * @throws Exception 如果解密失败
     */
    public static String decrypt(String ciphertext, String version) throws Exception {
        SecretKey key = deriveKey(version);

        // 解码 Base64
        byte[] combined = Base64.decode(ciphertext, Base64.NO_WRAP);

        // 提取 IV
        byte[] iv = Arrays.copyOfRange(combined, 0, GCM_IV_LENGTH);

        // 提取实际密文
        byte[] encrypted = Arrays.copyOfRange(combined, GCM_IV_LENGTH, combined.length);

        // 初始化解密器
        Cipher cipher = Cipher.getInstance(TRANSFORMATION);
        GCMParameterSpec parameterSpec = new GCMParameterSpec(GCM_TAG_LENGTH, iv);
        cipher.init(Cipher.DECRYPT_MODE, key, parameterSpec);

        // 解密
        byte[] plaintext = cipher.doFinal(encrypted);
        return new String(plaintext, StandardCharsets.UTF_8);
    }

    /**
     * 解密服务器额外加密的字段
     * 使用服务器提供的会话密钥
     *
     * 安全说明：
     * - 服务器额外加密用于高度敏感字段（api_key, secret_key）
     * - 当前实现：抛出异常提醒尚未实现
     * - 未来实现：使用服务器协商的会话密钥进行二次解密
     *
     * @param ciphertext 服务器额外加密的密文
     * @param sessionKey 会话密钥（从服务器获取），当前未使用
     * @return 解密后的明文
     * @throws UnsupportedOperationException 当前功能尚未实现
     */
    public static String decryptServerEncrypted(String ciphertext, String sessionKey) {
        // 安全优先原则：对于未实现的服务器解密，抛出异常而不是静默失败
        Log.e(TAG, "Server decryption not yet implemented - cannot decrypt SERVER_ENC: fields");
        throw new UnsupportedOperationException(
            "Server-side decryption not yet implemented. " +
            "SERVER_ENC: fields require server session key negotiation. " +
            "Please use regular encryption for now."
        );
    }

    /**
     * 批量解密 JSON 对象中的加密字段
     *
     * 安全说明：
     * - 任何字段解密失败都会抛出异常，不会静默失败
     * - SERVER_ENC: 前缀的字段需要服务器解密（当前抛出异常）
     *
     * @param encryptedObj 加密的 JSON 对象
     * @param version 版本号
     * @return 解密后的 JSON 对象
     * @throws RuntimeException 如果任何字段解密失败
     */
    public static com.google.gson.JsonObject decryptJsonObject(
            com.google.gson.JsonObject encryptedObj,
            String version) {
        com.google.gson.JsonObject decrypted = new com.google.gson.JsonObject();

        for (String key : encryptedObj.keySet()) {
            try {
                String encryptedValue = encryptedObj.get(key).getAsString();

                // 检查是否为服务器额外加密
                if (encryptedValue.startsWith("SERVER_ENC:")) {
                    String serverEncrypted = encryptedValue.substring("SERVER_ENC:".length());
                    String decryptedValue = decryptServerEncrypted(serverEncrypted, null);
                    decrypted.addProperty(key, decryptedValue);
                } else {
                    // 常规版本号加密
                    String decryptedValue = decrypt(encryptedValue, version);
                    decrypted.addProperty(key, decryptedValue);
                }
            } catch (Exception e) {
                // 不再静默失败 - 记录详细错误并重新抛出
                Log.e(TAG, "CRITICAL: Failed to decrypt field '" + key + "'. " +
                    "This may indicate: 1) Corrupted config, 2) Wrong version, 3) Tampering detected", e);
                throw new RuntimeException("Failed to decrypt config field: " + key, e);
            }
        }

        return decrypted;
    }
}
